demonstrations/tutorial_mol_geo_opt.py

```diff
@@ -0,0 +1,427 @@
+    grad = [qml.ExpvalCost(circuit, obs, dev)(params) for obs in grad_h]
+    return np.array(grad)
+
+##############################################################################
+# Optimization of the molecular geometry
+# --------------------------------------
+#
+# Finally, we proceed to minimize our cost function to find the ground state equilibrium
+# geometry of the :math:`\mathrm{H}_3^+` molecule. As a reminder, 
+# the circuit parameters and the nuclear coordinates will be jointly optimized at
+# each optimization step. This approach does not require a nested VQE
+# optimizations of the circuit parameters for each set of the nuclear coordinates.
+#
+# We start by defining the classical optimizers:
+
+opt_theta = qml.GradientDescentOptimizer(stepsize=0.4)
+opt_x = qml.GradientDescentOptimizer(stepsize=0.8)
+
+##############################################################################
+# Next, we initialize the circuit parameters :math:`\theta`. The angles
+# :math:`\theta_1` and :math:`\theta_2` are set to zero to approximate the
+# the initial state :math:`\vert\Psi(\theta_1, \theta_2)\rangle` 
+# as the Hartree-Fock state.
+
+theta = [0.0, 0.0]
+
+##############################################################################
+# On the other hand, the initial set of nuclear coordinates :math:`x`, defined at
+# the beginning of the tutorial, was computed classically within the Hartree-Fock
+# approximation using the GAMESS program [#ref_gamess]_. This is a natural choice
+# for the starting geometry that we are aiming to improve due to the electronic
+# correlation effects included in the trial state :math:`\vert\Psi(\theta)\rangle`.
+#
+# We carry out the optimization over a maximum of 100 steps. 
+# The circuit parameters and the nuclear coordinates are optimized until the
+# maximum component of the nuclear gradient :math:`\nabla_x g(\theta,x)` is
+# less than or equal to :math:`10^{-5}` Hartree/Bohr. Typically, this is the
+# convergence criterion used for optimizing molecular geometries in
+# quantum chemistry simulations.
+
+# store the values of the cost function
+energy = []
+
+# store the values of the H-H bond length
+bond_length = []
+
+# Factor to convert from Bohrs to Angstroms
+bohr_angs = 0.529177210903
+
+for n in range(100):
+
+    # Optimize the circuit parameters
+    theta = opt_theta.step(partial(cost, x=x), theta)
+
+    # Optimize the nuclear coordinates
+    grad_fn = partial(grad_x, params=theta)
+    x = opt_x.step(partial(cost, params=theta), x, grad_fn=grad_fn)
+
+    energy.append(cost(theta, x))
+    bond_length.append(np.linalg.norm(x[0:3] - x[3:6]) * bohr_angs)
+
+    if n % 2 == 0:
+        print(f"Iteration = {n},  Energy = {energy[-1]:.8f} Ha,  bond length = {bond_length[-1]:.5f} A")
+
+    # Check maximum component of the nuclear gradient
+    if np.max(grad_fn(x)) <= 1e-05:
+        break
+
+print("\n" "Final value of the ground-state energy = {:.8f} Ha".format(energy[-1]))
```

Comment:
```suggestion
print("\n" f"Final value of the ground-state energy = {energy[-1]:.8f} Ha")
```
Best to use f-strings here too, as you do below. I think this should work, but good to double-check